### ISSUE TITLE
mapping `oa_fetch()` freezes after around 4000 requests

### ISSUE THREAD

**Issue by JiaZhang42** [ropensci/openalexR#266]
When `oa_fetch()` was applied to a tibble column (~ 19k rows) through the `purrr::map()` function, it freezes after a while every time (progress bar showing 19%). Doing it with a small tibble (10 rows) doesn't run into any problem. 

I checked OpenAlex's rate limit and the number of requests is within that limit.

As the problem only occurs on a large scale, it's not easy to include a minimal reprex. The following code gives an idea of what happened.

``` r
library(tidyverse)
library(openalexR)
#> Thank you for using openalexR!
#> To acknowledge our work, please cite the package by calling `citation("openalexR")`.
#> To suppress this message, add `openalexR.message = suppressed` to your .Renviron file.

publications_tb
#> # A tibble: 19,117 × 2
#>   title_clean                                                                year
#>   <chr>                                                                     <dbl>
#> 1 a bias free assessment of the hedge fund industry a new evaluation of to…  2023
#> 2 a note on the valuation of asset management firms                          2017
#> 3 are hedge funds exploiting climate concerns                                2023
#> 4 effect of regulatory constraints on fund performance new evidence from u…  2015
#> 5 frictional diversification costs evidence from a panel of fund of hedge …  2019
#> 6 hedge fund performance are stylized facts sensitive to which database on…  2019
#> 7 hedge fund performance end of an era                                       2021
#> 8 hedge fund portfolio selection with fund characteristics                   2021
#> 9 hedge fund selection best athletes or portfolio gains                        NA
#> 10 hedge fund tail risk performance and hedging mechanisms                    2015
#> # ℹ 19,107 more rows
#> # ℹ Use `print(n = ...)` to see more rows

get_doi_from_title_year <- function(title, year){
  if (is.na(year)) {
    oa_fetch(
      entity = 'works', 
      display_name.search = title
    ) %>% 
      slice_head(n=1) %>% 
      select(id, title, display_name, author, publication_year, so, url, doi) %>% 
      suppressWarnings()
  } else {
    oa_fetch(
      entity = 'works', 
      display_name.search = title, 
      publication_year = year
    ) %>% 
      slice_head(n=1) %>% 
      select(id, title, display_name, author, publication_year, so, url, doi) %>% 
      suppressWarnings()
  }
}

publications_tb_openalex <- publications_tb %>% 
  mutate(
    openalex = map2(title_clean, year, possibly(get_doi_from_title_year, NA), .progress = T)
  )
#> ■■■■■■■                           19% |  ETA: 13h
```

Does anyone have any idea what the issue could be? Thanks!

**Comment by rkrug** [comment]
The problem is likely the conversion into a tibble of the results. This needs a large amount of memory, and the time needed grows (due to processing) in my experience non linear. I would suggest to try to download as list and convert afterwards. Cheers,RainerSent from my iPhoneOn 12 Aug 2024, at 04:45, Jia Zhang ***@***.***> wrote:﻿
When oa_fetch() was applied to a tibble column (~ 19k rows) through the purrr::map() function, it freezes after a while every time (progress bar showing 19%). Doing it with a small tibble (10 rows) doesn't run into any problem.
I checked OpenAlex's rate limit and the number of requests is within that limit.
As the problem only occurs on a large scale, it's not easy to include a minimal reprex. The following code gives an idea of what happened.
library(tidyverse)
library(openalexR)
#> Thank you for using openalexR!
#> To acknowledge our work, please cite the package by calling `citation("openalexR")`.
#> To suppress this message, add `openalexR.message = suppressed` to your .Renviron file.

publications_tb
#> # A tibble: 19,117 × 2
#>   title_clean                                                                year
#>   <chr>                                                                     <dbl>
#> 1 a bias free assessment of the hedge fund industry a new evaluation of to…  2023
#> 2 a note on the valuation of asset management firms                          2017
#> 3 are hedge funds exploiting climate concerns                                2023
#> 4 effect of regulatory constraints on fund performance new evidence from u…  2015
#> 5 frictional diversification costs evidence from a panel of fund of hedge …  2019
#> 6 hedge fund performance are stylized facts sensitive to which database on…  2019
#> 7 hedge fund performance end of an era                                       2021
#> 8 hedge fund portfolio selection with fund characteristics                   2021
#> 9 hedge fund selection best athletes or portfolio gains                        NA
#> 10 hedge fund tail risk performance and hedging mechanisms                    2015
#> # ℹ 19,107 more rows
#> # ℹ Use `print(n = ...)` to see more rows

get_doi_from_title_year <- function(title, year){
  if (is.na(year)) {
    oa_fetch(
      entity = 'works', 
      display_name.search = title
    ) %>% 
      slice_head(n=1) %>% 
      select(id, title, display_name, author, publication_year, so, url, doi) %>% 
      suppressWarnings()
  } else {
    oa_fetch(
      entity = 'works', 
      display_name.search = title, 
      publication_year = year
    ) %>% 
      slice_head(n=1) %>% 
      select(id, title, display_name, author, publication_year, so, url, doi) %>% 
      suppressWarnings()
  }
}

publications_tb_openalex <- publications_tb %>% 
  mutate(
    openalex = map2(title_clean, year, possibly(get_doi_from_title_year, NA), .progress = T)
  )
#> ■■■■■■■                           19% |  ETA: 13h
Does anyone have any idea what the issue could be? Thanks!

**Comment by JiaZhang42** [comment]
> The problem is likely the conversion into a tibble of the results. This needs a large amount of memory, and the time needed grows (due to processing) in my experience non linear. I would suggest to try to download as list and convert afterwards. Cheers,Rainer>

Thank you, Rainer! I checked the activity manager and R was indeed using a lot of memory. It's very likely where the problem is. I'll try your advice out.

**Comment by rkrug** [comment]
Set verbiisity to TRUE and see if it always stops at the same work. Try to narrow it down to the work and try it from there. If this does not help, I have a custom implementation of oa_request() which saves each retrieved page. You can find it at openalexR/R/oa_fetch.R at IPBES.openalexR · IPBES-Data/openalexRgithub.comPlease ferl free to use it or any ideas from it. Cheers Sent from my iPhoneOn 12 Aug 2024, at 18:11, Jia Zhang ***@***.***> wrote:﻿
I tried to fetch JSON results rather than tibble. Surprisingly, it froze again at the 19% progress bar.
I guess it could be

something goes wrong after any 19%*nrow requests;
something went wrong at the 19% quantile of my requests.

So I sliced out data between the 18% to 20% quantiles and tested it again. It got stuck too.
It turned out some extremely short titles caused the problem. They by default would result in a huge amount of results returned, exploding the memory use. Setting options = list( sample = 1) in oa_fetch() or oa_query() would solve this problem.

**Comment by rkrug** [comment]
Perfect. Sent from my iPhoneOn 12 Aug 2024, at 18:11, Jia Zhang ***@***.***> wrote:﻿
I tried to fetch JSON results rather than tibble. Surprisingly, it froze again at the 19% progress bar.
I guess it could be

something goes wrong after any 19%*nrow requests;
something went wrong at the 19% quantile of my requests.

So I sliced out data between the 18% to 20% quantiles and tested it again. It got stuck too.
It turned out some extremely short titles caused the problem. They by default would result in a huge amount of results returned, exploding the memory use. Setting options = list( sample = 1) in oa_fetch() or oa_query() would solve this problem.
